clients/drcachesim/tracer/instr_counter.cpp

```diff
@@ -72,28 +78,73 @@ static uint64 instr_count;
 // but to avoid the complexity of different instrumentation we need a threshold.
 #define DELAY_FOREVER_THRESHOLD (1024 * 1024 * 1024)
 
-static std::atomic<bool> reached_trace_after_instrs;
+static std::atomic<bool> reached_trace_after_instrs(false);
 std::atomic<uint64> retrace_start_timestamp;
 
+static std::atomic<uint64> instr_intervals_idx(0);
+static std::vector<uint64> trace_for_instrs_list;
+static std::vector<uint64> no_trace_for_instrs_list;
+
+void
+increment_instr_intervals_index()
+{
+    if (!trace_for_instrs_list.empty())
+        instr_intervals_idx.fetch_add(1);
+}
+
+uint64
+get_trace_after_instrs_value()
+{
+    if (op_trace_after_instrs.get_value() > 0)
+        return op_trace_after_instrs.get_value();
+    if (!no_trace_for_instrs_list.empty())
+        return no_trace_for_instrs_list[0];
+    return 0;
+}
+
+uint64
+get_trace_for_instrs_value()
+{
+    if (op_trace_for_instrs.get_value() > 0)
+        return op_trace_for_instrs.get_value();
+    if (!trace_for_instrs_list.empty()) {
+        uint64 i = instr_intervals_idx.load(std::memory_order_acquire);
```

Comment:
Add out-of-bounds check

Reply:
Ok, should we call `FATAL()` if that happens?

Reply:
I was gonna suggest assert, but those are disabled in non-debug builds. If we know it's an out of bounds access, better to fail with FATAL even on opt yes.

Reply:
Done.